diff --git a/hw/xquartz/pbproxy/Makefile.am b/hw/xquartz/pbproxy/Makefile.am
index 9429ea25d..e2d52f259 100644
--- a/hw/xquartz/pbproxy/Makefile.am
+++ b/hw/xquartz/pbproxy/Makefile.am
@@ -1,7 +1,7 @@
 AM_CPPFLAGS=-DBUNDLE_ID_PREFIX=\"$(BUNDLE_ID_PREFIX)\"
 
 AM_CFLAGS=$(XPBPROXY_CFLAGS)
-AM_OBJCFLAGS=$(XPBPROXY_CFLAGS)
+AM_OBJCFLAGS=$(XPBPROXY_CFLAGS) -fobjc-exceptions
 
 noinst_LTLIBRARIES = libxpbproxy.la
 libxpbproxy_la_SOURCES = \
diff --git a/hw/xquartz/xpr/x-list.h b/hw/xquartz/xpr/x-list.h
index 28385fd99..87cccb48e 100644
--- a/hw/xquartz/xpr/x-list.h
+++ b/hw/xquartz/xpr/x-list.h
@@ -44,6 +44,10 @@ struct x_list_struct {
 #define X_PFX(x) x_ ## x
 #endif
 
+#ifndef __private_extern__
+#define __private_extern__  extern
+#endif
+
 #ifndef X_EXTERN
 #define X_EXTERN __private_extern__
 #endif
