diff --git bench/bench.c bench/bench.c
index 39d6aeb..3bc617f 100644
--- bench/bench.c
+++ bench/bench.c
@@ -121,6 +121,21 @@ double get_usec_chunk(blosc_timestamp_t last, blosc_timestamp_t current, int nit
   return elapsed_usecs / (double)(niter * nchunks);
 }
 
+#if defined __APPLE__ && MAC_OS_X_VERSION_MAX_ALLOWED < 1060
+#include <errno.h>
+int posix_memalign(void **memptr, size_t alignment, size_t size)
+{
+   void* mem = 0;
+   mem = valloc(size);
+   if (mem == 0)
+       return ENOMEM;
+   else {
+       *memptr = mem;
+       return 0;
+   }
+}
+#endif
+
 /* Define posix_memalign for Windows */
 #if defined(_WIN32)
 #include <malloc.h>
