--- ./projects/compiler-rt/cmake/Modules/CompilerRTDarwinUtils.cmake.orig	2017-05-26 21:35:22.000000000 -0700
+++ ./projects/compiler-rt/cmake/Modules/CompilerRTDarwinUtils.cmake	2017-05-26 21:35:49.000000000 -0700
@@ -37,7 +37,7 @@
   else()
     # If auto-detecting fails, fall back to a default set
     message(WARNING "Detecting supported architectures from 'ld -v' failed. Returning default set.")
-    set(ARCHES "i386;x86_64;armv7;armv7s;arm64")
+    set(ARCHES "i386;x86_64;ppc;ppc64;armv7;armv7s;arm64")
   endif()
   
   set(${output_var} ${ARCHES} PARENT_SCOPE)
